skyportal/handlers/api/annotation.py

```diff
@@ -0,0 +1,263 @@
+                              type: integer
+                              description: New annotation ID
+        """
+        data = self.get_json()
+        obj_id = data.get("obj_id")
+        origin = data.get("origin")
+
+        if obj_id is None:
+            return self.error("Missing required field `obj_id`")
+
+        if origin is None or origin == '':
+            return self.error('Missing required field `origin`')
+
+        annotation_data = data.get("data")
+
+        # Ensure user/token has access to parent source
+        obj = Source.get_obj_if_owned_by(obj_id, self.current_user)
+        user_accessible_group_ids = [g.id for g in self.current_user.accessible_groups]
+        user_accessible_filter_ids = [
+            filtr.id
+            for g in self.current_user.accessible_groups
+            for filtr in g.filters
+            if g.filters is not None
+        ]
+        group_ids = [int(id) for id in data.pop("group_ids", user_accessible_group_ids)]
+        group_ids = set(group_ids).intersection(user_accessible_group_ids)
+        if not group_ids:
+            return self.error(
+                f"Invalid group IDs field ({group_ids}): "
+                "You must provide one or more valid group IDs."
+            )
+
+        # Only post to groups source/candidate is actually associated with
+        if DBSession().query(Candidate).filter(Candidate.obj_id == obj_id).all():
+            candidate_group_ids = [
+                f.group_id
+                for f in (
+                    DBSession()
+                    .query(Filter)
+                    .filter(Filter.id.in_(user_accessible_filter_ids))
+                    .filter(
+                        Filter.id.in_(
+                            DBSession()
+                            .query(Candidate.filter_id)
+                            .filter(Candidate.obj_id == obj_id)
+                        )
+                    )
+                    .all()
+                )
+            ]
+        else:
+            candidate_group_ids = []
+        matching_sources = (
+            DBSession().query(Source).filter(Source.obj_id == obj_id).all()
+        )
+        if matching_sources:
+            source_group_ids = [source.group_id for source in matching_sources]
+        else:
+            source_group_ids = []
+        group_ids = set(group_ids).intersection(candidate_group_ids + source_group_ids)
+        if not group_ids:
+            return self.error("Obj is not associated with any of the specified groups.")
+
+        groups = Group.query.filter(Group.id.in_(group_ids)).all()
+
+        author = self.associated_user_object
+        annotation = Annotation(
+            data=annotation_data,
+            obj_id=obj_id,
+            origin=origin,
+            author=author,
+            groups=groups,
+        )
+
+        DBSession().add(annotation)
+        if 'redshift' in annotation_data:
+            obj.redshift = annotation_data['redshift']
```

Comment:
```suggestion
```

We dont want to handle redshifts here. There is going to be a separate handler for redshifts based on the classifications API

Reply:
I know, but this functionality exists in the current Comment code, so I thought it would be good to pull it out only after we have something else instead.